rubygem/lib/zeus.rb

```diff
@@ -192,28 +194,18 @@ def report_error_to_master(local, error)
       local.write str
     end
 
-    def run_action(socket, identifier, feature_pipe_w)
+    def run_action(socket, identifier)
       # Now we run the action and report its success/fail status to the master.
-      features, err = Zeus::LoadTracking.features_loaded_by do
-        plan.after_fork unless identifier == :boot
-        plan.send(identifier)
-      end
-
-      if err
-        # If we received an error, report features to the master syncronously.
-        # We need to do this before reporting the error to the master
-        # otherwise it will kill us before we can report features.
-        begin
-          notify_features(feature_pipe_w, features)
-          feature_pipe_w.close
-        ensure
-          report_error_to_master(socket, err)
+      begin
+        Zeus::LoadTracking.track_features_loaded_by do
+          plan.after_fork unless identifier == :boot
+          plan.send(identifier)
         end
-      else
-        # If we booted successfully, report features in a new thread
```

Comment:
Getting rid of this is a somewhat significant behavior change. We'll now wait until after notifying features until marking the process as running and allowing it to fork. Given that forking and threading interact really badly, I think this is better for stability. Notifying features for every Ruby file in our codebase only takes ~100ms so this isn't a big delay anyway.